File: Android/src/app/src/main/java/com/google/ai/edge/gallery/customtasks/speech/SpeechAudio.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ai.edge.gallery.customtasks.speech

import android.annotation.SuppressLint
import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioManager
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaRecorder
import android.util.Log
import com.google.ai.edge.gallery.data.CategoryInfo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.withContext

private const val TAG = "AGSpeechAudio"

/** The sample rate (Hz) shared by the speech models bundled with the Speech tasks. */
const val SPEECH_SAMPLE_RATE = 16000

/**
 * The home-screen category (tab) that groups the speech-related custom tasks (Text to Speech and
 * Speech to Text).
 */
val SpeechCategory = CategoryInfo(id = "speech", label = "Speech")

/**
 * Plays back PCM audio produced by a TTS model.
 *
 * Samples are expected to be mono floats in the [-1, 1] range, which is exactly what
 * `sherpa-onnx`'s `GeneratedAudio.samples` provides. Playback runs on a background thread so the
 * caller is never blocked.
 */
class AudioPlayer {
  private var track: AudioTrack? = null
  private var writerThread: Thread? = null

  fun play(samples: FloatArray, sampleRate: Int) {
    stop()
    if (samples.isEmpty()) {
      return
    }

    val minBufferSize =
      AudioTrack.getMinBufferSize(
        sampleRate,
        AudioFormat.CHANNEL_OUT_MONO,
        AudioFormat.ENCODING_PCM_FLOAT,
      )
    val bufferSize = if (minBufferSize > 0) minBufferSize else sampleRate * 4

    val newTrack =
      AudioTrack(
        AudioAttributes.Builder()
          .setUsage(AudioAttributes.USAGE_MEDIA)
          .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
          .build(),
        AudioFormat.Builder()
          .setEncoding(AudioFormat.ENCODING_PCM_FLOAT)
          .setSampleRate(sampleRate)
          .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
          .build(),
        bufferSize,
        AudioTrack.MODE_STREAM,
        AudioManager.AUDIO_SESSION_ID_GENERATE,
      )
    track = newTrack
    newTrack.play()

    writerThread =
      Thread {
          try {
            var offset = 0
            while (offset < samples.size && !Thread.currentThread().isInterrupted) {
              val written =
                newTrack.write(
                  samples,
                  offset,
                  samples.size - offset,
                  AudioTrack.WRITE_BLOCKING,
                )
              if (written <= 0) {
                break
              }
              offset += written
            }
          } catch (e: Exception) {
            Log.w(TAG, "Audio playback interrupted", e)
          }
        }
        .also { it.start() }
  }

  /**
   * Plays [samples] and suspends until playback finishes (or the calling coroutine is cancelled).
   *
   * Unlike [play] (fire-and-forget), this is meant for **sequential streaming playback**: a caller
   * can `playToCompletion(sentence1); playToCompletion(sentence2); …` and each clip plays fully
   * before the next begins. Cancelling the coroutine (e.g. on barge-in) stops playback promptly.
   */
  suspend fun playToCompletion(samples: FloatArray, sampleRate: Int) {
    stop()
    if (samples.isEmpty()) {
      return
    }
    withContext(Dispatchers.IO) {
      val minBufferSize =
        AudioTrack.getMinBufferSize(
          sampleRate,
          AudioFormat.CHANNEL_OUT_MONO,
          AudioFormat.ENCODING_PCM_FLOAT,
        )
      val bufferSize = if (minBufferSize > 0) minBufferSize else sampleRate * 4
      val t =
        AudioTrack(
          AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_MEDIA)
            .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
            .build(),
          AudioFormat.Builder()
            .setEncoding(AudioFormat.ENCODING_PCM_FLOAT)
            .setSampleRate(sampleRate)
            .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
            .build(),
          bufferSize,
          AudioTrack.MODE_STREAM,
          AudioManager.AUDIO_SESSION_ID_GENERATE,
        )
      track = t
      try {
        t.play()
        var offset = 0
        while (offset < samples.size && isActive) {
          val written =
            t.write(samples, offset, samples.size - offset, AudioTrack.WRITE_BLOCKING)
          if (written <= 0) {
            break
          }
          offset += written
        }
        // The last write returns once the data is buffered; wait until the playback head has
        // actually reached the end so the clip isn't cut off before the next one starts.
        while (isActive && t.playbackHeadPosition < samples.size) {
          delay(20)
        }
      } catch (e: Exception) {
        Log.w(TAG, "Sequential audio playback interrupted", e)
      } finally {
        try {
          t.pause()
          t.flush()
          t.stop()
        } catch (e: Exception) {
          Log.w(TAG, "Failed to stop AudioTrack", e)
        }
        t.release()
        if (track === t) {
          track = null
        }
      }
    }
  }

  fun stop() {
    writerThread?.interrupt()
    writerThread = null
    track?.let { t ->
      try {
        t.pause()
        t.flush()
        t.stop()
      } catch (e: Exception) {
        Log.w(TAG, "Failed to stop AudioTrack", e)
      }
      t.release()
    }
    track = null
  }
}

/**
 * Records microphone audio as mono float samples in the [-1, 1] range, ready to be fed directly
 * into a `sherpa-onnx` recognizer.
 *
 * The caller is responsible for holding the `RECORD_AUDIO` runtime permission before calling
 * [start].
 */
class AudioRecorder(private val sampleRate: Int = SPEECH_SAMPLE_RATE) {
  @Volatile private var recording = false
  private var recordThread: Thread? = null
  private val chunks = mutableListOf<FloatArray>()

  @SuppressLint("MissingPermission")
  fun start() {
    if (recording) {
      return
    }
    synchronized(chunks) { chunks.clear() }

    val minBufferSize =
      AudioRecord.getMinBufferSize(
        sampleRate,
        AudioFormat.CHANNEL_IN_MONO,
        AudioFormat.ENCODING_PCM_16BIT,
      )
    val bufferSize = if (minBufferSize > 0) minBufferSize * 2 else sampleRate * 2

    val recorder =
      AudioRecord(
        MediaRecorder.AudioSource.VOICE_RECOGNITION,
        sampleRate,
        AudioFormat.CHANNEL_IN_MONO,
        AudioFormat.ENCODING_PCM_16BIT,
        bufferSize,
      )
    if (recorder.state != AudioRecord.STATE_INITIALIZED) {
      recorder.release()
      throw IllegalStateException("Failed to initialize AudioRecord")
    }

    recording = true
    recorder.startRecording()
    recordThread =
      Thread {
          val buffer = ShortArray(bufferSize / 2)
          try {
            while (recording) {
              val read = recorder.read(buffer, 0, buffer.size)
              if (read > 0) {
                val floats = FloatArray(read) { buffer[it] / 32768.0f }
                synchronized(chunks) { chunks.add(floats) }
              }
            }
          } finally {
            try {
              recorder.stop()
            } catch (e: Exception) {
              Log.w(TAG, "Failed to stop AudioRecord", e)
            }
            recorder.release()
          }
        }
        .also { it.start() }
  }

  /** Stops recording and returns all captured samples concatenated into a single array. */
  fun stop(): FloatArray {
    if (!recording) {
      return FloatArray(0)
    }
    recording = false
    recordThread?.join(2000)
    recordThread = null

    synchronized(chunks) {
      val total = chunks.sumOf { it.size }
      val out = FloatArray(total)
      var offset = 0
      for (chunk in chunks) {
        chunk.copyInto(out, offset)
        offset += chunk.size
      }
      chunks.clear()
      return out
    }
  }

  fun isRecording(): Boolean = recording
}

File: iOS/Gallery/CustomTasks/Speech/SpeechAudio.swift
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License"); see LICENSE.
//
// Port of customtasks/speech/SpeechAudio.kt

import AVFoundation
import Foundation

/// The sample rate (Hz) shared by the speech models bundled with the Speech tasks.
let SPEECH_SAMPLE_RATE: Int = 16_000

/// The home-screen category that groups the speech-related custom tasks
/// (Text to Speech and Speech to Text).
let SpeechCategory = CategoryInfo(id: "speech", label: "Speech")

// MARK: - AudioPlayer

/// Plays back PCM audio produced by a TTS engine.
///
/// Samples are expected to be mono floats in the [-1, 1] range.
/// Playback runs off the main thread so the caller is never blocked.
final class AudioPlayer {
  private var player: AVAudioPlayer?
  private var playerTask: _Concurrency.Task<Void, Never>?

  /// Fire-and-forget playback. If something is already playing it is stopped first.
  func play(samples: [Float], sampleRate: Int) {
    stop()
    guard !samples.isEmpty else { return }
    playerTask = _Concurrency.Task.detached { [weak self] in
      await self?.playInternal(samples: samples, sampleRate: sampleRate)
    }
  }

  /// Suspends until playback finishes (or the calling Task is cancelled).
  /// Suitable for sequential sentence-by-sentence playback.
  func playToCompletion(samples: [Float], sampleRate: Int) async {
    stop()
    guard !samples.isEmpty else { return }
    await playInternal(samples: samples, sampleRate: sampleRate)
  }

  private func playInternal(samples: [Float], sampleRate: Int) async {
    // Build a 16-bit PCM WAV buffer in memory.
    guard let data = buildWav(samples: samples, sampleRate: sampleRate) else { return }
    do {
      let p = try AVAudioPlayer(data: data)
      player = p
      p.prepareToPlay()
      p.play()
      // Poll until the player finishes or the Task is cancelled.
      while p.isPlaying {
        if _Concurrency.Task.isCancelled { p.stop(); break }
        try? await _Concurrency.Task.sleep(nanoseconds: 20_000_000)  // 20 ms
      }
    } catch {
      // Non-fatal; the UI already shows an error if synthesis itself failed.
    }
  }

  func stop() {
    playerTask?.cancel()
    playerTask = nil
    player?.stop()
    player = nil
  }

  // MARK: WAV builder

  /// Encodes [-1,1] float samples as a 16-bit PCM mono WAV file in memory.
  private func buildWav(samples: [Float], sampleRate: Int) -> Data? {
    let numSamples = samples.count
    let byteCount = numSamples * 2
    let headerSize = 44
    var data = Data(count: headerSize + byteCount)

    func write32LE(_ value: UInt32, at offset: Int) {
      data[offset]     = UInt8(value & 0xFF)
      data[offset + 1] = UInt8((value >> 8) & 0xFF)
      data[offset + 2] = UInt8((value >> 16) & 0xFF)
      data[offset + 3] = UInt8((value >> 24) & 0xFF)
    }
    func write16LE(_ value: UInt16, at offset: Int) {
      data[offset]     = UInt8(value & 0xFF)
      data[offset + 1] = UInt8((value >> 8) & 0xFF)
    }

    // RIFF header
    data[0...3]   = Data("RIFF".utf8)
    write32LE(UInt32(36 + byteCount), at: 4)
    data[8...11]  = Data("WAVE".utf8)
    data[12...15] = Data("fmt ".utf8)
    write32LE(16, at: 16)          // PCM chunk size
    write16LE(1, at: 20)           // PCM format
    write16LE(1, at: 22)           // mono
    write32LE(UInt32(sampleRate), at: 24)
    write32LE(UInt32(sampleRate * 2), at: 28)  // byte rate
    write16LE(2, at: 32)           // block align
    write16LE(16, at: 34)          // bits per sample
    data[36...39] = Data("data".utf8)
    write32LE(UInt32(byteCount), at: 40)

    // PCM samples
    var offset = headerSize
    for s in samples {
      let clamped = max(-1, min(1, s))
      let raw = Int16(clamped * 32767)
      data[offset]     = UInt8(bitPattern: Int8(raw & 0x7F) )
      data[offset]     = UInt8(raw & 0xFF)
      data[offset + 1] = UInt8((raw >> 8) & 0xFF)
      offset += 2
    }
    return data
  }
}

// MARK: - AudioRecorder

/// Records microphone audio as mono float samples in the [-1, 1] range at
/// `SPEECH_SAMPLE_RATE` Hz, suitable for on-device STT engines.
///
/// The caller is responsible for holding the `NSMicrophoneUsageDescription`
/// / microphone permission before calling `start()`.
///
/// NOTE: On Android this used `AudioRecord` (PCM_16BIT). On iOS we use
/// `AVAudioEngine` with a tap on the input node, converting the hardware
/// buffer's native format to 16 kHz mono PCM on the fly.
/// To integrate a real on-device sherpa-onnx / CoreML STT engine, feed
/// the `[Float]` returned by `stop()` directly into the recognizer's
/// `acceptWaveform(samples:sampleRate:)` equivalent.
final class AudioRecorder {
  private let sampleRate: Int
  private var engine: AVAudioEngine?
  private var chunks: [[Float]] = []
  private let lock = NSLock()
  private(set) var isRecording = false

  init(sampleRate: Int = SPEECH_SAMPLE_RATE) {
    self.sampleRate = sampleRate
  }

  /// Starts capturing microphone audio. Throws if the engine can't start.
  func start() throws {
    guard !isRecording else { return }
    lock.lock(); chunks.removeAll(); lock.unlock()

    let e = AVAudioEngine()
    engine = e

    let inputNode = e.inputNode
    let hwFormat = inputNode.outputFormat(forBus: 0)
    guard let targetFormat = AVAudioFormat(
      commonFormat: .pcmFormatFloat32,
      sampleRate: Double(sampleRate),
      channels: 1,
      interleaved: false)
    else {
      throw RecordingError.formatUnsupported
    }

    // NOTE: `AVAudioConverter` translates from the hardware's native sample
    // rate / channel layout to 16 kHz mono float. Replace this path with a
    // sherpa-onnx streaming recognizer tap when integrating on-device ASR.
    guard let converter = AVAudioConverter(from: hwFormat, to: targetFormat) else {
      throw RecordingError.formatUnsupported
    }

    let bufferSize = AVAudioFrameCount(hwFormat.sampleRate * 0.1)  // 100 ms chunks
    inputNode.installTap(onBus: 0, bufferSize: bufferSize, format: hwFormat) {
      [weak self] buffer, _ in
      guard let self else { return }
      let ratio = targetFormat.sampleRate / hwFormat.sampleRate
      let outFrames = AVAudioFrameCount(Double(buffer.frameLength) * ratio + 0.5)
      guard let outBuffer = AVAudioPCMBuffer(pcmFormat: targetFormat, frameCapacity: outFrames) else { return }
      var error: NSError?
      converter.convert(to: outBuffer, error: &error) { _, status in
        status.pointee = .haveData
        return buffer
      }
      if let channelData = outBuffer.floatChannelData?[0] {
        let count = Int(outBuffer.frameLength)
        let floats = Array(UnsafeBufferPointer(start: channelData, count: count))
        self.lock.lock(); self.chunks.append(floats); self.lock.unlock()
      }
    }

    try e.start()
    isRecording = true
  }

  /// Stops recording and returns all captured samples concatenated.
  func stop() -> [Float] {
    guard isRecording else { return [] }
    isRecording = false
    engine?.inputNode.removeTap(onBus: 0)
    engine?.stop()
    engine = nil

    lock.lock(); defer { lock.unlock() }
    var out: [Float] = []
    for c in chunks { out.append(contentsOf: c) }
    chunks.removeAll()
    return out
  }

  enum RecordingError: Error {
    case formatUnsupported
    case engineStart(Error)
  }
}
